add options to list focused tests

vilinski opened the pull request:
following PR to enhancement #392  

Proposed changes and usage:
- `--list-tests` - lists all tests as before
- `--list-tests focused` - lists focused tests only
- `--list-tests pending focused` - lists pending and focused tests only
- added a "state char" before test name (N, P, F). It is hidden as before, if none of listed tests are focused or pending (not the case in Expecto.Tests)

```
P pending_test.case.one 
F focused_test.case.two
N normal_test.case.three 
```

haf commented on Expecto/Expecto.fs:
```diff
@@ -402,6 +408,8 @@ module Tests =
   type FillFromArgsResult =
     | ArgsRun of ExpectoConfig
     | ArgsList of ExpectoConfig
```
Wouldn't these be better placed in the config?

haf commented on Expecto/Expecto.fs:
```diff
@@ -472,6 +482,10 @@ module Tests =
             Seq.fold (fun s a -> foldCLIArgumentToConfig a s) baseConfig cliArguments
           if List.contains List_Tests cliArguments then
             ArgsList config
```
Inside this DU, that is?

vilinski commented:
Hope I understand you right. 
- added a field into ExpectoConfig
- changed the option `--list-tests` to optionally receive list of test states, added tests
- added entries into release notes for this and previous PR

also please check wording, my english is not that strong

haf commented on Expecto/Expecto.Impl.fs:
```diff
@@ -494,6 +494,8 @@ module Impl =
       /// An optional filter function. Useful if you only would
       /// like to run a subset of all the tests defined in your assembly.
       filter   : Test -> Test
+      /// List tests of specified state
```
What does this mean? Is this a configuration flag?

haf replied:
So this is a configuration flag to show only all tests that have one or two or three of the specified states?

vilinski replied:
the proposed usage would be:

- `--list-tests` - lists all tests as before
- `--list-tests focused` - lists focused tests only
- `--list-tests pending focused` - lists pending and focused tests

but I'm of course open for any alternatives.

haf commented on Expecto/Expecto.fs:
```diff
@@ -387,9 +384,7 @@ module Tests =
-      "--list-tests", "Don't run tests, but prints out list of tests instead.", Args.none List_Tests
-      "--list-focused-tests", "Don't run tests, but prints out list of focused tests instead.", Args.none List_Focused_Tests
-      "--list-pending-tests", "Don't run tests, but prints out list of pending tests instead.", Args.none List_Pending_Tests
+      "--list-tests", "Don't run tests, but prints out list of tests instead. Lists only tests with specified state(s), or all tests if not.", Args.list (Args.focusState) List_Tests
```
Perhaps we should make it a `--summary-state` or something similar to the existing "display only" mnemonics?

vilinski replied:
you mean let `--list-tests` alone but add `--summary-state` to display states?

vilinski replied:
what you think about just add states (or/and maybe colors) to `--list-tests`?

```
P pending_test.case.one 
F focused_test.case.two
N normal_test.case.three 
```

haf replied:
I didn't realise we had this flag from before; let's leave it alone; if it can work without an argument (as a default showing everything) that's good. Also good would be your latest suggestion about displaying state next to each test.

vilinski replied:
Sorry for delay, I have added a "state char" before test names. Updated PR description. 👎 or 👍 ?

haf commented on RELEASE_NOTES.md:
```diff
@@ -1,3 +1,7 @@
+#### 9.0.3 - 2020-07-27
```
👍

haf commented:
Build needs fixing and there's a `printf`: wouldn't it be better to go through the logger like everywhere else? How available is the logger where you've made your change?

vilinski commented:
> Build needs fixing

Oh, I see, there is a bug in NuGet/paket which I'm also struggling with. https://github.com/NuGet/Home/issues/9786
Workaround with commenting out a line in `paket.Restore.targets` didn't work. No clue how to fix this until it's done in NuGet

> and there's a `printf`: wouldn't it be better to go through the logger like everywhere else? How available is the logger where you've made your change?

`printfn` is a shadowed function which was used here before. `Impl.logger` is also awailable from `listTests` function. But I'm failed to figure out how to use it right. Is `logger.info` a right thing to do here? My last commit results in no output at all. 
actually it does if I append printfn after log or `Async.RunSynchronously`. Seems it's not get flushed before program ends.

haf commented:
@vilinski You can do an Async.RunSynchronously for now; the most important bit is that it goes through the logger so we can quelch output, should we wish to. I'm not too happy on shadowed stdlib functions.

vilinski commented:
`Async.RunSynchronously` and `logWithAck` doesn't help. it only works with `printfn ""` at end ...
An alternative would be to return the string from `listTests` instead of logging it.
also all other occurences of shadowed printfn are in the calling function `runTestsWithArgsAndCancel`.

haf commented:
The solution here is not to add a "fake" printf, but to hunt down why the logger's flush doesn't work. I'm afraid I don't have time to spare right now, but I would very much appreciate if you feel up for it.

vilinski commented:
tried to look into code and debug. `Monitor.Enter` and `Monitor.Exit` calls in progress printer are looking plausible for me.

however noticed that the `Logger.log` and `Logger.logWithAck` are identical in all leaf implementations `LiterateConsoleTarget`, `TextWriterTarget`, `OutputWindowTarget`

I have added a flush into `LiterateConsoleTarget`:

```
$ dotnet run --framework netcoreapp3.1 --  --list-tests pending
[00:37:38 INF] 
P findFastest.find fastest sleep
P all focused tests/ignored by attribute
P performance.same function is faster should fail
P findFastest.find fastest lo
P findFastest.find fastest hi
 <Expecto>

```

Comparet to no output at all for the above command it's a progress. 
But can you please confirm this is the right way to do? If not I'm afraid I can't understand how to fix this either.

haf commented on Expecto/Expecto.fs:
```diff
@@ -484,9 +485,37 @@ module Tests =
+    Printf.bprintf result "\n"
+
+    logger.logWithAck Info (
+      Message.eventX "{result}"
```
Better to make the event the result IMO;

vilinski replied:
please explain, `Message.eventX result` without fields?

haf commented:
I think we can live with that flush call; if all other tests are passing. What happens if you disable progress printing? Flush still works, but is a no-op?

vilinski commented:
> I think we can live with that flush call; if all other tests are passing. 

No failing tests in `Expecto.Tests`.  What about other two `logWithAck`s? Do they also need a flush?

> What happens if you disable progress printing? 

`--no-spinner` option? I see the same output without progress spinner in either cases, maybe because there is just one log call for `--list-tests`

Still don't know hot to fix the build though

haf commented:
/rebase

haf commented:
Thanks @vilinski ! 🎊
